src/cartridges/adyen_controllers_changes/app_storefront_controllers_changes/cartridge/controllers/PaymentInstruments.js

```diff
@@ -0,0 +1,326 @@
+ */
+function add(clearForm) {
+    var paymentForm = app.getForm('paymentinstruments');
+    if (clearForm !== false) {
+        paymentForm.clear();
+    }
+    paymentForm.get('creditcards.newcreditcard.type').setOptions(dw.order.PaymentMgr
+        .getPaymentMethod(dw.order.PaymentInstrument.METHOD_CREDIT_CARD).activePaymentCards.iterator());
+
+    app.getView({
+        ContinueURL: URLUtils.https('PaymentInstruments-PaymentForm'),
+        SessionData: JSON.stringify(getSessionData()),
+    }).render('account/payment/paymentinstrumentdetails');
+}
+
+/**
+ * Form handler for the paymentinstruments form. Handles the following actions:
+ * - __create__ - calls the {@link module:controllers/PaymentInstruments~create|create} function to create a payment instrument
+ * and redirects to {@link module:controllers/PaymentInstruments~list|list}. If the
+ * creation fails, calls the {@link module:controllers/PaymentInstruments~add|add} function with a clearform value of false.
+ * - __error__ - calls the {@link module:controllers/PaymentInstruments~add|add} function with a clearform value of false.
+ */
+function handlePaymentForm() {
+    var paymentForm = app.getForm('paymentinstruments');
+    paymentForm.handleAction({
+        create: function () {
+            if (!create()) {
+                response.redirect(URLUtils.https('PaymentInstruments-List', 'error', 'AuthorisationFailed'));
+                return;
+            } else {
+                response.redirect(URLUtils.https('PaymentInstruments-List'));
+            }
+        },
+        error: function () {
+            add(false);
+        }
+    });
+}
+/**
+ * Saves a  customer credit card payment instrument.
+ * @param {Object} params
+ * @param {dw.customer.CustomerPaymentInstrument} params.PaymentInstrument - credit card object.
+ * @param {dw.web.FormGroup} params.CreditCardFormFields - new credit card form.
+ */
+function save(params) {
+    var saveCustomerCreditCard = require('app_storefront_core/cartridge/scripts/checkout/SaveCustomerCreditCard');
+    var result = saveCustomerCreditCard.save(params);
+    if (result === PIPELET_ERROR) {
+        throw new Error('Problem saving credit card');
+    }
+}
+
+/**
+ * Creates a new payment instrument. Verifies the credit card and checks if it is a duplicate of
+ * a card already in the current customer's payment instruments. In a transaction, the function
+ * attempts to save the credit card to the customer's payment instruments. If a duplicate card was
+ * detected, the original card is removed after the new card is created. If the card cannot be created
+ * successfully, the transaction is rolled back. Whether successful or not, the paymentinstruments
+ * form is cleared.
+ *
+ * @transaction
+ * @return {boolean} true if the credit card can be verified, false otherwise
+ */
+function create() {
+    if(getAdyenPaymentInstruments()) {
```

Comment:
Should we add the code comments to indicate what is the custom Adyen code?

Reply:
@maassenbas alright I wrapped the methods that involve changes with our comment tags